trackers/run_tracker.py

```diff
@@ -72,26 +85,62 @@ def run(global_config_file, dash_schema_file, pipelines, run_id=1):
                         status = func(subject_dir, session_id, run_id)
                         # print(f"task_name: {name}, status: {status}")
                         _df.loc[bids_id,name] = status
-                        _df.loc[bids_id,"pipeline_starttime"] = get_start_time(subject_dir)
-                        _df.loc[bids_id,"pipeline_endtime"] = UNAVAILABLE # TODO
+                    _df.loc[bids_id,"pipeline_starttime"] = get_start_time(subject_dir)
+                    _df.loc[bids_id,"pipeline_endtime"] = get_end_time(subject_dir)
+                    n_participants += 1
                 else:
                     # print(f"Pipeline output not found for bids_id: {bids_id}, session: {session}")
                     for name in status_check_dict.keys():                    
                         _df.loc[bids_id,name] = UNAVAILABLE
-                        _df.loc[bids_id,"pipeline_starttime"] = UNAVAILABLE
-                        _df.loc[bids_id,"pipeline_endtime"] = UNAVAILABLE
+                    _df.loc[bids_id,"pipeline_starttime"] = UNAVAILABLE
+                    _df.loc[bids_id,"pipeline_endtime"] = UNAVAILABLE
+
+                # don't check all participants if testing
+                if testing and n_participants > N_TESTING:
+                    break
+
+            proc_status_session_dfs.append(_df)
 
-            proc_status_dfs.append(_df)
+        # new rows for this pipeline
+        pipeline_proc_status_df = pd.concat(proc_status_session_dfs, axis='index').reset_index(drop=True)
 
-    proc_status_df = pd.concat(proc_status_dfs, axis='index')
+        # add old rows from other pipelines and sort for consistent order
+        proc_status_df: pd.DataFrame = pd.concat([old_proc_status_df, pipeline_proc_status_df], axis='index')
```

Comment:
It's for type annotation so that my code editor knows that this is a `pd.DataFrame` and shows the right available methods (since `pd.concat()` can return `pd.Series` in some cases). It doesn't actually enforce that `proc_status_df` must be a dataframe

Reply:
Oh, neat, I didn’t realize you could use type annotation “in the wild” like this 🤓

Reply:
For some things, mypy will actually error if you do not declare the type of new variables.

One that is also handy when you start using mypy and addting type hints for constants:
https://mypy.readthedocs.io/en/stable/final_attrs.html